### ISSUE TITLE
Prices, Currencies, and Filters

### ISSUE THREAD

**Issue by plaidscreen** [getkart/kart#20]
Ok how about this:
Everything is stored and calculated in whole numbers. Currency is installed as an extension, and the Currrency extension automatically filters the output displayed to the user. The theme shouldn't be doing any math, so I think it would be safe if the `cart()->item()->price()` function actually spat it out with the currency already converted.

Examples: Say you have a product with a price of 100.
If a USD currency extension is installed, it would need to show $1.00 to the user. So it hooks into the price filter:
```PHP
Event::on("Kart/item/price", function($price) {
  return "$" . number_format($price / 100, 2);
});
```
If bitcoin is installed, it would have to convert the price by multiplying the item price by the store minimum (since bitcoins can be infinitesimal)
```PHP
Event::on("Kart/item/price", function($price) {
  return "Ƀ" . $price * c::get("Kart/currency/bitcoin/minimum", 1000);
  // let's just assume extensions are using standard kirby config for now
});
```
And then in our core, we have
```PHP
public static function price() {
  // ...
  return Event::filter("Kart/item/price",$this->price);
}
```

**Comment by plaidscreen** [comment]
One more thing I just thought of:
I feel like `->price()` should be returning the whole number. Perhaps themes should use a different function, like `->formatted_price()`. Alternatively, to get the whole number, one could use `->price` (to get the variable directly), and `->price()` (a function) to get the formatted version

Another thing to think about are stores that have multiple currencies. Some stores charge a completely different price in one currency that doesn't evenly convert to another. Apple, for instance, sells an iPad for $499 USD. But in Australia, it's $699 AUD. Those prices are not equivalent, at current exchange rates, it would cost significantly more to buy an iPad in Australian dollars. So, Apple's store database obviously isn't just saving one price and converting it later, they're storing two prices. I have an idea how we could tackle this, but I'll save it for another day.

**Comment by plaidpowered** [comment]
Something about the bitcoin solution is not sitting right with me. Interested to hear some ideas.

**Comment by benseitz** [comment]
I really like your idea of having special currency extensions (let's not forget about decimal point vs decimal comma) - and having them as integers for the minimum quantity (e.g. cents)
Since I don't have any experience with bitcoin what so ever, I can't even say if it feels right.. :D Maybe we should ask someone who's more into them? ;)

**Comment by plaidpowered** [comment]
Man I just don't know how to do this one.

That said, here's another idea: Products have an array of prices.

The blueprint's price field is a structure with two fields, `value` and `currency`. `value` is a number (which for all major currencies should be an integer, but for cryptocurrencies, a float), and currency is a select. The select would fill its options with a list of the currently installed currencies (which of course, isn't possible yet without panel hooks)
Example:
```
prices:
  label: Prices
  type: structure
  entry: > 
    {{value}} {{currency}}
  fields:
    value:
      type: number
      label: Price
    currency:
      type: select
      label: currency
      options: query
      query: (don't know if possible yet)
```

And for a blueprint-free version:
```PHP
$kart->cart()->add(array(
  "title" => "Something",
  "ID" => "product_id",
  "qty" => 1,
  "price" => array(
    "usd"=>100,
    "aud"=>110,
    "bc"=>0.004
  )
);
```
During checkout, the user selects their currency manually, or it's automatically determined through their location. Products that don't have a price available for the currency of choice will not be calculated or charged, but instead removed from the final order.

So a currency extension would have to provide the following filters at minimum:
- Function to output a human readable format
- Function that adds the currency key and label to a global list of available currencies


This price stuff is so messy! But definitely better to figure out before we start seriously writing code.

**Comment by benseitz** [comment]
I agree price and taxes are really messy...

Your thoughts sound very good. having int for traditional currency and float for cyber currency is a simple but awesome idea!

And there also should be a function to convert currency depending an exchange rate (if a shop owner prefers this over defining individual prices)

**Comment by plaidpowered** [comment]
@benseitz Perhaps "Exchange" would be a separate extension. It could still pull from the list of installed currencies, but the Exchange extension would keep track of conversion rates and modify the price displayed to the customer. Written smartly enough, it could be a "last step" fall back, if no price was available for the customer's selected currency, the Exchange extension tries to figure it out.

**Comment by distantnative** [comment]
I agree that exchange screams more like optional extension, cause there is even more that you can do wrong with displaying calculated exchange prices to customers that don't match their bank statements in the end ;)

And I like your concept, especially given that a lot of payment gateways want exactly this pair of information, currency identifier and amount in cents (or whatever). :+1:

**Comment by benseitz** [comment]
:+1: 
Of course as an optional plugin, but still a possibility :)

**Comment by plaidpowered** [comment]
I suppose at this point the de facto assumption is that _everything_ we talk about is an option extension. :smiley:

I was even flirting with the idea of making Price an extension too. I think that might be going too far though.

**Comment by benseitz** [comment]
:smile:
